src/coreclr/pal/src/thread/context.cpp

```diff
@@ -1880,3 +1880,29 @@ DBG_FlushInstructionCache(
 #endif
     return TRUE;
 }
+
+#ifdef HOST_AMD64
+CONTEXT& CONTEXT::operator=(const CONTEXT& ctx)
+{
+    size_t copySize;
+    if (ctx.ContextFlags & CONTEXT_XSTATE & CONTEXT_AREA_MASK)
+    {
+        if ((ctx.XStateFeaturesMask & XSTATE_MASK_AVX512) == XSTATE_MASK_AVX512)
+        {
+            copySize = sizeof(CONTEXT);
+        }
+        else
+        {
+            copySize = offsetof(CONTEXT, KMask0);
+        }
+    }
+    else
+    {
+        copySize = offsetof(CONTEXT, Ymm0H);
```

Comment:
Shouldn't `ymm` only be there if we have the `XSTATE`?

That is, by default the `CONTEXT` struct only tracks the general purpose registers + the XMM_SAVE_AREA. `AVX` (which includes the upper-bits for the YMM state), `AVX512_KMASK`, `AVX512_ZMM_H`, and `AVX512_ZMM` are all separate flags for `XSTATE`.

For the proper `CONTEXT` in Windows, we'd normally get the `ContextSize` once up front (from `InitializeContext`) and then be able to simply `memcpy` using the returned `contextSize`. We'd then locate the subset of these feature areas when required via the `LocateXStateFeature` API and determine which were available via `GetXStateFeaturesMask`.

I wonder if modeling it more this way to match would make it cheaper/simpler (or potentially not having the PAL in the way for this perf sensitive scenario at all, particularly since there are far more optimal ways to represent this info if we did that).

Reply:
I am a bit confused - when XSTATE is not set, we copy everything upto to Ymm0H, so only the xmms are there. 
Anyways, we should not copy the XStateFeaturesMask and XStateReserved0 in this case.

As for making the CONTEXT XState handling closer to how Windows does it (including the CopyContext / InitializeContext functions) might make sense, but I am not sure if it would really simplify anything.

In fact, we use the copying of the CONTEXT without using the real size at most places on Windows too (just assigning CONTEXT by value). I guess that those places just don't care about the xstate at all, so copying the base is sufficient. It just makes the target context incorrect (xstate indicating there is some xstate, but the copy not having any), so if some new code started to use such copies not having this on mind, subtle issues could occur.

Reply:
> I am a bit confused - when XSTATE is not set, we copy everything upto to Ymm0H, so only the xmms are there.
Anyways, we should not copy the XStateFeaturesMask and XStateReserved0 in this case

Ah right, I misread this as copying including `Ymm0H` and not copy up to `Ymm0H`, my mistake

> In fact, we use the copying of the CONTEXT without using the real size at most places on Windows too (just assigning CONTEXT by value). I guess that those places just don't care about the xstate at all, so copying the base is sufficient. It just makes the target context incorrect (xstate indicating there is some xstate, but the copy not having any), so if some new code started to use such copies not having this on mind, subtle issues could occur.

Right. It'd be nice if we could be more explicit about these things to help avoid such subtle issues (not necessarily in this PR, just in general for the future).